### ISSUE TITLE
Avoid copying Pandas DataFrames for Python bindings when possible

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#1706]
This issue is specific to the Python bindings and came out of the review of #1700 and the related issue #1694.  mlpack Python bindings support Pandas DataFrames, as in the following snippet:

```
import pandas as pd
import mlpack

d = pd.read_csv("my_data.csv")
output = mlpack.pca(input=d, verbose=True)
```

The Pandas DataFrame `d` gets automatically converted to a numpy matrix, and then the underlying data pointer of that numpy matrix is passed to the binding itself.

The code flow is that the method `mlpack.pca()` (or any other mlpack Python method) will call `to_matrix()` with an input matrix, which will then "convert" whatever was given to a numpy `ndarray`.  (Later, outside the scope of `to_matrix()`, the returned `ndarray` is converted to an Armadillo matrix without copying.  We don't need to worry about those details for this issue.)

Currently the implementation of `to_matrix()` operates basically by checking if what was given to `to_matrix()` is already an `ndarray` that is row-major (i.e. the flag `C_CONTIGUOUS` is set).  If that's true (and we are not explicitly told to copy the matrix), we return the matrix that was passed with no copy or modification.  But if the input to `to_matrix()` is *not* an `ndarray`, we end up copying no matter what.

So what we should aim to look into this issue is the following:

1. Does Pandas store an underlying DataFrame in the same memory layout as an `ndarray`?  If so, then...
2. Adapt `to_matrix()` to handle Pandas DataFrames (and possibly Pandas series too) without copying, when possible.

This issue can make a big difference, because for large datasets even just copying the memory itself can have significant cost.

**Comment by mirraaj** [comment]
@rcurtin @zoq  Could you please provide the link where function ```to_matrix``` is defined?

**Comment by rcurtin** [comment]
Oh!  Forgot to do that.  It's in `src/mlpack/bindings/python/mlpack/matrix_utils.py`.  Thanks for pointing that out. :)

**Comment by mirraaj** [comment]
@rcurtin Let me see if I could contribute to this issue. Thanks

**Comment by Yashwants19** [comment]
Hi @rcurtin please review #1711

**Comment by rcurtin** [comment]
@Yashwants19 I saw the PR, I just haven't had a chance to respond to it yet.  Don't worry---I get an email for all activity, so I will respond when I'm able to.

**Comment by g-laz77** [comment]
Is this issue still open?

**Comment by abernauer** [comment]
If this issue still needs to be addressed up for helping.

**Comment by rcurtin** [comment]
Ah, sorry, I should have closed this---with #1711 merged, I think this is fixed.  Thanks everyone for the help. :+1:
